src/Microsoft.VisualBasic/tests/StringsTests.cs

```diff
@@ -104,6 +104,99 @@ public void ChrW_CharCodeOutOfRange_ThrowsArgumentException(int charCode)
+        [InlineData(new string[] { "bab" }, "a", new string[] { "bab" }, new string[] { })]
+        [InlineData(new string[] { "b" }, "a", new string[] { }, new string[] { "b" })]
+        [InlineData(new string[] { "a" }, "ab", new string[] { }, new string[] { "a" })]
+        [InlineData(new string[] { "ab" }, "ab", new string[] { "ab" }, new string[] { })]
+        public void Filter_SingleElement(string[] source, string match, string[] includeExpected, string[] excludeExpected)
+        {
+            Assert.Equal(includeExpected, Strings.Filter(source, match, Include: true));
+            Assert.Equal(excludeExpected, Strings.Filter(source, match, Include: false));
+        }
+
+        [Theory]
+        [InlineData(new string[] { "A" }, "a", new string[] { }, new string[] { "A" })]
+        public void Filter_SingleElement_BinaryCompare(string[] source, string match, string[] includeExpected, string[] excludeExpected)
+        {
+            Assert.Equal(includeExpected, Strings.Filter(source, match, Include: true, Compare: CompareMethod.Binary));
+            Assert.Equal(excludeExpected, Strings.Filter(source, match, Include: false, Compare: CompareMethod.Binary));
+        }
+
+        [Theory]
+        [InlineData(new string[] { "A" }, "a", new string[] { "A" }, new string[] { })]
+        public void Filter_SingleElement_TextCompare(string[] source, string match, string[] includeExpected, string[] excludeExpected)
+        {
+            Assert.Equal(includeExpected, Strings.Filter(source, match, Include: true, Compare: CompareMethod.Text));
+            Assert.Equal(excludeExpected, Strings.Filter(source, match, Include: false, Compare: CompareMethod.Text));
+        }
+
+        [Theory]
+        [InlineData(new string[] { "a", "a" }, "a", new string[] { "a", "a" }, new string[] { })]
+        [InlineData(new string[] { "a", "b" }, "a", new string[] { "a" }, new string[] { "b" })]
+        [InlineData(new string[] { "b", "a" }, "a", new string[] { "a" }, new string[] { "b" })]
+        [InlineData(new string[] { "b", "b" }, "a", new string[] { }, new string[] { "b", "b" })]
+        public void Filter_MultipleElements(string[] source, string match, string[] includeExpected, string[] excludeExpected)
+        {
+            Assert.Equal(includeExpected, Strings.Filter(source, match, Include: true));
+            Assert.Equal(excludeExpected, Strings.Filter(source, match, Include: false));
+        }
+
+        [Fact]
+        public void Filter_Objects_WhenObjectCannotBeConvertedToString_ThrowsArgumentOutOfRangeException()
+        {
+            var source = new object[] { typeof(object) };
+            var match = "a";
+
+            var exception = Assert.Throws<ArgumentException>(() => Strings.Filter(source, match));
+            var expectedMessage = "Argument 'Source' cannot be converted to type 'String'.\r\nParameter name: Source";
+            Assert.Equal(expectedMessage, exception.Message);
```

Comment:
Why not test for the exact message? By testing the message I discovered that the message of the exceptions thrown in `Strings.Left`, `Strings.Mid`, … are faulty (bug should be created). They don't format the message string, which results in the text '...Argument '{0}' must be...'.

@danmosemsft, what's your opinion on testing the exception properties?

Reply:
We don't test exception strings because (1) they are stripped in the .NET Native (UWP) build and (2) if and when we localize them, we'll have to revisit all the tests. We can test the ParamName on ArgumentExceptions though. See https://github.com/dotnet/corefx/blob/master/src/CoreFx.Private.TestUtilities/src/System/AssertExtensions.cs . 

> don't format the message string, which results in the text '...Argument '{0}' must be...'.

This is an easy mistake to make and it would be nice to catch it. You could try modifying AssertExtensions.Throws where they check the ParamName to check that ParamName is also found within Message. You would have to run all the tests in the tree though because some code might possibly not include the parameter in the message.